Library/Homebrew/unpack_strategy.rb

```diff
@@ -0,0 +1,338 @@
+class UnpackStrategy
+  # length of the longest regex (currently TarUnpackStrategy)
+  MAX_MAGIC_NUMBER_LENGTH = 262
+  private_constant :MAX_MAGIC_NUMBER_LENGTH
+
+  def self.strategies
+    @strategies ||= [
+      JarUnpackStrategy,
+      ZipUnpackStrategy,
+      XarUnpackStrategy,
+      CompressUnpackStrategy,
+      TarUnpackStrategy,
+      GzipUnpackStrategy,
+      Bzip2UnpackStrategy,
+      XzUnpackStrategy,
+      LzipUnpackStrategy,
+      GitUnpackStrategy,
+      MercurialUnpackStrategy,
+      SubversionUnpackStrategy,
+      CvsUnpackStrategy,
+      FossilUnpackStrategy,
+      BazaarUnpackStrategy,
+      P7ZipUnpackStrategy,
+      RarUnpackStrategy,
+      LhaUnpackStrategy,
+    ].freeze
+  end
+  private_class_method :strategies
+
+  def self.detect(path, ref_type: nil, ref: nil)
+    magic_number = if path.directory?
+      ""
+    else
+      File.binread(path, MAX_MAGIC_NUMBER_LENGTH) || ""
```

Comment:
Given the `""` fallback in both cases I wonder if:

```ruby
magic_number = File.binread(path, MAX_MAGIC_NUMBER_LENGTH) if path.directory?
magic_number ||= ""
```

or similar might be better?

Reply:
I tried some variations, but I like this version because it is more explicit.